src/evict/evict_page.c

```diff
@@ -248,7 +248,7 @@ __evict_page_dirty_update(WT_SESSION_IMPL *session, WT_REF *ref, bool closing)
 	switch (mod->rec_result) {
 	case WT_PM_REC_EMPTY:				/* Page is empty */
 		/* Discard the parent's address. */
-		__wt_ref_free_addr(session, ref);
+		__wt_ref_addr_free(session, ref);
```

Comment:
I think you're correct (nice catch!).

The code that frees the block is:
```
static int
__rec_write_wrapup(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_PAGE *page)
{
        ...
        switch (mod->rec_result) {
        case 0: /*
                 * The page has never been reconciled before, free the original
                 * address blocks (if any).  The "if any" is for empty trees
                 * created when a new tree is opened or previously deleted pages
                 * instantiated in memory.
                 *
                 * The exception is root pages are never tracked or free'd, they
                 * are checkpoints, and must be explicitly dropped.
                 */
                if (__wt_ref_is_root(ref))
                        break;
                WT_RET(__wt_ref_block_free(session, ref));
                break;
```
and it should happen on the first reconciliation, from which point on, `ref->addr` should stay NULL. I can't think of any reason the path to `WT_PM_REC_MULTIBLOCK` is any different from the two cases you noted, and it doesn't free `ref->addr`, which lends weight to the argument.

It's a tricky change, though: for now, we could assert at the top of `__evict_page_dirty_update` that `ref->addr == NULL`?